StakingERC721/contracts/StakingERC721.sol

```diff
@@ -0,0 +1,371 @@
+      revert WaitToFinish();
+    }
+    rewardsDuration = duration;
+    emit RewardsDurationUpdated(duration);
+  }
+
+  /**
+   * @notice Stakes user's NFTs
+   * @dev Callable only when unpaused
+   * @param tokenIds The tokenIds of the NFTs which will be staked
+   */
+  function stake(
+    uint256[] memory tokenIds
+  ) public nonReentrant whenNotPaused updateReward(msg.sender) {
+    if (tokenIds.length == 0) {
+      revert NoTokenIds();
+    }
+
+    uint256 amount;
+    uint256 length = tokenIds.length;
+    for (uint256 i = 0; i < length; ++i) {
+      // Transfer user's NFTs to the staking contract
+      stakingToken.safeTransferFrom(msg.sender, address(this), tokenIds[i]);
+      // Increment the amount which will be staked
+      ++amount;
+      // Save who is the staker/depositor of the token
+      stakedAssets[tokenIds[i]] = msg.sender;
+    }
+    _stake(amount);
+    emit Staked(msg.sender, amount, tokenIds);
+  }
+
+  /**
+   * @notice Unstake staked user's NFTs
+   * @dev Callable only when unpaused
+   * @param tokenIds The tokenIds of the NFTs which will be withdrawn
+   */
+  function unstake(
+    uint256[] memory tokenIds
+  ) public nonReentrant whenNotPaused updateReward(msg.sender) {
+    if (tokenIds.length == 0) {
+      revert NoTokenIds();
+    }
+
+    uint256 amount;
+    uint256 length = tokenIds.length;
+    for (uint256 i = 0; i < length; ++i) {
+      // Check if the user who withdraws is the owner
+      if (stakedAssets[tokenIds[i]] != msg.sender) {
+        revert NotTokenOwner();
+      }
+      // Transfer NFTs back to the owner
+      stakingToken.safeTransferFrom(address(this), msg.sender, tokenIds[i]);
+      // Increment the amount which will be withdrawn
+      ++amount;
+      // Cleanup stakedAssets for the current tokenId
+      stakedAssets[tokenIds[i]] = address(0);
+    }
+    _unstake(amount);
+
+    emit Unstaked(msg.sender, amount, tokenIds);
+  }
+
+  /**
+   * @notice Claim rewards tokens, callable only when unpaused.
+   * @dev Callable only when unpaused
+   */
+  function claim() public nonReentrant whenNotPaused updateReward(msg.sender) {
+    uint256 reward = rewards[msg.sender];
+    if (reward > 0) {
+      rewards[msg.sender] = 0;
+      rewardsToken.safeTransfer(msg.sender, reward);
+      emit Claimed(msg.sender, reward);
+    }
+  }
+
+  /**
+   * @notice Claim rewards tokens and unstake staked amount.
+   * @dev Callable only when unpaused.
+   */
+  function exit(uint256[] memory tokenIds) public whenNotPaused {
+    unstake(tokenIds);
+    claim();
+  }
+
+  /**
+   * @notice Fund rewards tokens and re-calculate rewards rate.
+   * Rewards rate will be calculated again from remaining distributable tokens
+   * and awarding rewards tokens.
+   * @dev Callable only when unpaused and by owner
+   * @param reward Funding rewards token amount
+   */
+  function fund(
+    uint256 reward
+  ) public onlyOwner whenNotPaused updateReward(address(0)) {
```

Comment:
We can use external to reduce gas fee, or maybe include this in interface

Reply:
I don't think that include `fund` function in the interface, this should be callable only by owner and not by contract.